datafusion/common/src/tree_node.rs

```diff
@@ -0,0 +1,261 @@
+                )))
+            }
+        };
+
+        for child in self.get_children() {
+            child.collect(op)?;
+        }
+
+        Ok(())
+    }
+
+    /// Visit the tree node using the given [TreeNodeVisitor]
+    /// It performs a depth first walk of an node and its children.
+    ///
+    /// For an node tree such as
+    /// ```text
+    /// ParentNode
+    ///    left: ChildNode1
+    ///    right: ChildNode2
+    /// ```
+    ///
+    /// The nodes are visited using the following order
+    /// ```text
+    /// pre_visit(ParentNode)
+    /// pre_visit(ChildNode1)
+    /// post_visit(ChildNode1)
+    /// pre_visit(ChildNode2)
+    /// post_visit(ChildNode2)
+    /// post_visit(ParentNode)
+    /// ```
+    ///
+    /// If an Err result is returned, recursion is stopped immediately
+    ///
+    /// If [`Recursion::Stop`] is returned on a call to pre_visit, no
+    /// children of that node will be visited, nor is post_visit
+    /// called on that node
+    ///
+    /// If using the default [`post_visit`] with nothing to do, the [`collect`] should be preferred
+    fn visit<V: TreeNodeVisitor<N = Self>>(&self, visitor: &mut V) -> Result<Recursion> {
+        match visitor.pre_visit(self)? {
+            Recursion::Continue => {}
+            // If the recursion should stop, do not visit children
+            Recursion::Stop => return Ok(Recursion::Stop),
+            r => {
+                return Err(DataFusionError::Execution(format!(
+                    "Recursion {r:?} is not supported for collect_using"
+                )))
+            }
+        };
+
+        for child in self.get_children() {
+            match child.visit(visitor)? {
+                Recursion::Continue => {}
+                // If the recursion should stop, do not visit children
+                Recursion::Stop => return Ok(Recursion::Stop),
+                r => {
+                    return Err(DataFusionError::Execution(format!(
+                        "Recursion {r:?} is not supported for collect_using"
+                    )))
+                }
+            }
+        }
+
+        visitor.post_visit(self)
+    }
+
+    /// Convenience utils for writing optimizers rule: recursively apply the given `op` to the node tree.
+    /// When `op` does not apply to a given node, it is left unchanged.
+    /// The default tree traversal direction is transform_up(Postorder Traversal).
+    fn transform<F>(self, op: &F) -> Result<Self>
+    where
+        F: Fn(Self) -> Result<Option<Self>>,
+    {
+        self.transform_up(op)
+    }
+
+    /// Convenience utils for writing optimizers rule: recursively apply the given 'op' to the node and all of its
+    /// children(Preorder Traversal).
+    /// When the `op` does not apply to a given node, it is left unchanged.
+    fn transform_down<F>(self, op: &F) -> Result<Self>
+    where
+        F: Fn(Self) -> Result<Option<Self>>,
+    {
+        let node_cloned = self.clone();
```

Comment:
Can we remove cloning?

Reply:
Thanks for your comments. I'm afraid it's not feasible because of the constraint of `F: Fn(Self) -> Result<Option<Self>>,`

Reply:
F will consume the `Self`.

Reply:
I think this clone could be quite bad for performance -- as it means any code that uses this will effectively deep copy (recursively) the entire tree at each step (N^2 performance)

I think the root cause is the signature

```ryst
F: Fn(Self) -> Result<Option<Self>>,
```

Which I think is being used to signal if the node was changed or not

In order to avoid the clone, I think we need to use something other than `Option` so the original can be returned when the change is not made. 

How about something like

```rust
enum Transformed<T> {
  /// The item was transformed / rewritten somehow
  Yes(T),
  /// The item was not transformed
  No(T)
}

impl<T> From<Transformed<T>> for <T> {
  fn from(transformed: Transformed<T>) -> T {
    match transformed {
      Yes(t) => t,
      No(t) => t,
    }
}
  
```

And then this function would look something like

```rust
    fn transform_down<F>(self, op: &F) -> Result<Self>
    where
        F: Fn(Self) -> Result<Transformed<Self>>,
    {
        // this could be cleaner using the `From` impl
        let after_op = match op(node_cloned)? {
            Yes(value) => value,
            No(value) => value,
        };
        after_op.map_children(|node| node.transform_down(op))
    }
```

Reply:
I am particular concerned that this trait will be applied to `Expr`s (where as LogicalPlans and PhysicalPlans already have Arc's in them so the cost is not that high)

I would be willing to try and change the existing LogicalPlan optimizer (which I think might be causing  this problems) 


@jackwener  and I were just discussing a similar issue with https://github.com/apache/arrow-datafusion/pull/5705#discussion_r1146314877